### PR TITLE
token syncer: adds ping-token sub-command

### PR DESCRIPTION
**Pull request by shamsimam**
## Changes proposed in this PR

- adds health-check-token to waiter-api
- adds sub-command for pinging tokens on multiple clusters

## Why are we making these changes?

We would like the ability to ensure a token can be used to successfully launch services on multiple clusters.

### Example usage:

```
$ lein run -- ping-token -q 10000 test-healthy-1234 http://localhost:9091 http://localhost:9093
2018-08-27 17:06:56 INFO  token-syncer.main - command-line arguments: [ping-token -q 10000 test-healthy-1234 http://localhost:9091 http://localhost:9093]
2018-08-27 17:06:57 INFO  token-syncer.commands.ping - pinging token test-healthy-1234 on clusters: #{http://localhost:9093 http://localhost:9091}
2018-08-27 17:06:57 INFO  token-syncer.waiter - health-check-token test-healthy-1234 on http://localhost:9093 with queue timeout of 10000 ms
2018-08-27 17:06:57 INFO  token-syncer.waiter - token-syncer.569640a3-0a74-4f71-8e2c-734aca9f8ad7 making GET request to http://localhost:9093/token
2018-08-27 17:06:57 INFO  token-syncer.waiter - loading test-healthy-1234 on http://localhost:9093 responded with status 200 {content-type application/json, etag E-ed0d92d0c2b01f94fc813d03a4dc404d, x-cid token-syncer.569640a3-0a74-4f71-8e2c-734aca9f8ad7}
2018-08-27 17:06:57 INFO  token-syncer.waiter - health-check-token test-healthy-1234 on http://localhost:9093/status
2018-08-27 17:06:57 INFO  token-syncer.waiter - token-syncer.bf432023-7550-4e46-9fe5-c83d068cf5a7 making GET request to http://localhost:9093/status
2018-08-27 17:06:57 INFO  token-syncer.commands.ping - http://localhost:9093 health check response: {:body Hello World, :headers {..., content-length 11}, :status 200}
2018-08-27 17:06:57 INFO  token-syncer.waiter - health-check-token test-healthy-1234 on http://localhost:9091 with queue timeout of 10000 ms
2018-08-27 17:06:57 INFO  token-syncer.waiter - token-syncer.302433a6-345f-40c6-826f-826be97c3b48 making GET request to http://localhost:9091/token
2018-08-27 17:06:57 INFO  token-syncer.waiter - loading test-healthy-1234 on http://localhost:9091 responded with status 200 {content-type application/json, etag E-5aefd0ecd1261b5b4ee87883a3244de0, x-cid token-syncer.302433a6-345f-40c6-826f-826be97c3b48}
2018-08-27 17:06:57 INFO  token-syncer.waiter - health-check-token test-healthy-1234 on http://localhost:9091/status
2018-08-27 17:06:57 INFO  token-syncer.waiter - token-syncer.2e265000-daeb-4f77-aaf1-c8de8b498e4e making GET request to http://localhost:9091/status
2018-08-27 17:06:57 INFO  token-syncer.commands.ping - http://localhost:9091 health check response: {:body Hello World, :headers {..., content-length 11}, :status 200}
2018-08-27 17:06:57 INFO  token-syncer.commands.ping - {:details
 {"http://localhost:9093"
  {:exit-code 0,
   :message
   "successfully pinged token test-healthy-1234 on http://localhost:9093, reason: health check returned status code 200"},
  "http://localhost:9091"
  {:exit-code 0,
   :message
   "successfully pinged token test-healthy-1234 on http://localhost:9091, reason: health check returned status code 200"}},
 :exit-code 0,
 :message
 "pinging token test-healthy-1234 on [http://localhost:9093 http://localhost:9091] was successful",
 :token "test-healthy-1234"}
2018-08-27 17:06:57 INFO  token-syncer.main - token-syncer: ping-token: exiting with code 0

$ lein run -- ping-token -q 10000 test-fail-1234 http://localhost:9091 http://localhost:9093
2018-08-27 17:03:26 INFO  token-syncer.main - command-line arguments: [ping-token -q 10000 test-fail-1234 http://localhost:9091 http://localhost:9093]
2018-08-27 17:03:27 INFO  token-syncer.commands.ping - pinging token test-fail-1234 on clusters: #{http://localhost:9093 http://localhost:9091}
2018-08-27 17:03:27 INFO  token-syncer.waiter - health-check-token test-fail-1234 on http://localhost:9093 with queue timeout of 10000 ms
2018-08-27 17:03:27 INFO  token-syncer.waiter - token-syncer.ca10e9d9-4c26-477d-ba27-12ca1af5634a making GET request to http://localhost:9093/token
2018-08-27 17:03:27 INFO  token-syncer.waiter - loading test-fail-1234 on http://localhost:9093 responded with status 200 {content-type application/json, etag E-7890399a41a843c5c4cf84d33b567da4, x-cid token-syncer.ca10e9d9-4c26-477d-ba27-12ca1af5634a}
2018-08-27 17:03:27 INFO  token-syncer.waiter - health-check-token not performed {:cluster-url http://localhost:9093, :deleted nil, :health-check-url nil, :token test-fail-1234}
2018-08-27 17:03:27 INFO  token-syncer.commands.ping - http://localhost:9093 health check response: {:body , :headers nil, :status nil}
2018-08-27 17:03:27 INFO  token-syncer.waiter - health-check-token test-fail-1234 on http://localhost:9091 with queue timeout of 10000 ms
2018-08-27 17:03:27 INFO  token-syncer.waiter - token-syncer.1f559ea7-db51-49e0-9652-2c166c25cdc2 making GET request to http://localhost:9091/token
2018-08-27 17:03:27 INFO  token-syncer.waiter - loading test-fail-1234 on http://localhost:9091 responded with status 200 {content-type application/json, etag E-7890399a41a843c5c4cf84d33b567da4, x-cid token-syncer.1f559ea7-db51-49e0-9652-2c166c25cdc2}
2018-08-27 17:03:27 INFO  token-syncer.waiter - health-check-token not performed {:cluster-url http://localhost:9091, :deleted nil, :health-check-url nil, :token test-fail-1234}
2018-08-27 17:03:27 INFO  token-syncer.commands.ping - http://localhost:9091 health check response: {:body , :headers nil, :status nil}
2018-08-27 17:03:27 INFO  token-syncer.commands.ping - {:details
 {"http://localhost:9093"
  {:exit-code 1,
   :message
   "unable to ping token test-fail-1234 on http://localhost:9093, reason: health check returned status code "},
  "http://localhost:9091"
  {:exit-code 1,
   :message
   "unable to ping token test-fail-1234 on http://localhost:9091, reason: health check returned status code "}},
 :exit-code 2,
 :message
 "pinging token test-fail-1234 on [http://localhost:9093 http://localhost:9091] failed",
 :token
 "test-fail-1234"}
2018-08-27 17:03:27 ERROR token-syncer.main - token-syncer: ping-token: exiting with code 2
```

### CODE REVIEW

File: token-syncer/integration/token_syncer/basic_test.clj [hunk 1]
```diff
@@ -37,9 +44,14 @@
...
-    (is (System/getenv "WAITER_URIS"))
-    (is (> (count (waiter-urls)) 1))))
+    (testing "waiter cluster uris"
+      (log/info "env.WAITER_URIS" (System/getenv "WAITER_URIS"))
+      (is (System/getenv "WAITER_URIS"))
+      (is (> (count (waiter-urls)) 1)))
+
+    (testing "kitchen command"
+      (log/info "env.WAITER_TEST_KITCHEN_CMD" (System/getenv "WAITER_TEST_KITCHEN_CMD"))
+      (is (not (str/blank? (System/getenv "WAITER_TEST_KITCHEN_CMD")))))))
```
Comment (dposada) [inline token-syncer/integration/token_syncer/basic_test.clj]: I don't understand what these tests are accomplishing

File: token-syncer/src/token_syncer/commands/ping.clj [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+                  (log/error th "unable to ping token" token "on" cluster-url)
+                  {:exit-code 1
+                   :message (str "unable to ping token " token " on " cluster-url
+                                 ", reason: " (.getMessage th))})))
+            (set cluster-urls))
+          exit-code (reduce + 0 (->> cluster-url->result vals (map :exit-code)))]
+      {:details cluster-url->result
+       :exit-code exit-code
+       :message (str "pinging token " token " on " (-> cluster-urls vec println with-out-str str/trim)
+                     " " (if (zero? exit-code) "was successful" "failed"))})
```
Comment (dposada) [inline token-syncer/src/token_syncer/commands/ping.clj]: The "Successfully" message starts with a capitalized word, but the others don't

File: token-syncer/src/token_syncer/commands/ping.clj [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+                            {:body (str/trim (str body))
+                             :headers headers
+                             :status status})
+                  (if (and (integer? status) (<= 200 status 299))
+                    {:exit-code 0
+                     :message (str "Successfully pinged token " token " on " cluster-url
+                                   ", reason: health check returned status code " status)}
+                    {:exit-code 1
+                     :message (str "unable to ping token " token " on " cluster-url
+                                   ", reason: health check returned status code " status)}))
```
Comment (dposada) [inline token-syncer/src/token_syncer/commands/ping.clj]: Should we not return the token, cluster URL, and status as fields in the return map?

File: token-syncer/src/token_syncer/commands/ping.clj [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+                (catch Throwable th
+                  (log/error th "unable to ping token" token "on" cluster-url)
+                  {:exit-code 1
+                   :message (str "unable to ping token " token " on " cluster-url
+                                 ", reason: " (.getMessage th))})))
```
Comment (dposada) [inline token-syncer/src/token_syncer/commands/ping.clj]: Same question - why not include token, cluster URL, and the message in the return map?

File: token-syncer/src/token_syncer/commands/ping.clj (hunk 2 above)
Comment (dposada) [inline token-syncer/src/token_syncer/commands/ping.clj]: Maybe the token should be at this top-level in the return map?

File: token-syncer/src/token_syncer/waiter.clj [hunk 5]
```diff
@@ -195,3 +195,26 @@
...
+  "Performs health check on a token on a specific cluster."
+  [^HttpClient http-client cluster-url token]
+  (log/info "health-check-token" token "on" cluster-url)
+  (let [{:keys [description]} (load-token http-client cluster-url token)
+        {:strs [deleted health-check-url]} description]
+    (if (and (not deleted) (not (str/blank? health-check-url)))
+      (do
+        (log/info "health-check-token" token "on" (str cluster-url health-check-url))
+        (-> (make-http-request http-client (str cluster-url health-check-url)
+                               :headers {"x-waiter-queue-timeout" (-> 2 t/minutes t/in-millis)
```
Comment (dposada) [inline token-syncer/src/token_syncer/waiter.clj]: Should this be configurable?

File: token-syncer/src/token_syncer/waiter.clj [hunk 6]
```diff
@@ -195,3 +195,26 @@
...
+                               :method :get
+                               :query-params {})
+            (update :body async/<!!)))
+      (do
+        (log/info "health-check-token not performed"
+                  {:cluster-url cluster-url
+                   :deleted deleted
+                   :health-check-url health-check-url
+                   :token token})
+        (comment "return nil")))))
```
Comment (dposada) [inline token-syncer/src/token_syncer/waiter.clj]: Why the comment?

File: token-syncer/test/token_syncer/waiter_test.clj [hunk 7]
```diff
@@ -90,7 +90,7 @@
       (let [error (Exception. "exception from test")]
         (with-redefs [make-http-request (fn [in-http-client-wrapper in-endopint-url & in-options]
                                           (is (= http-client-wrapper in-http-client-wrapper))
-                                          (is (= (str test-cluster-url "/token")) in-endopint-url)
+                                          (is (= (str test-cluster-url "/tokens") in-endopint-url))
```
Comment (dposada) [inline token-syncer/test/token_syncer/waiter_test.clj]: Is this related to this change?

File: token-syncer/integration/token_syncer/basic_test.clj (hunk 1 above)
Comment (shamsimam) [reply token-syncer/integration/token_syncer/basic_test.clj]: These assertions are ensuring the integration tests are running with the required environment variables being present.

File: token-syncer/src/token_syncer/commands/ping.clj (hunk 3 above)
Comment (shamsimam) [reply token-syncer/src/token_syncer/commands/ping.clj]: I chose to log status and body instead of returning it as it is not used upstream in the caller.

The sub-command operators on a single token at a time, so I did not include the token in the map.

File: token-syncer/src/token_syncer/commands/ping.clj (hunk 4 above)
Comment (shamsimam) [reply token-syncer/src/token_syncer/commands/ping.clj]: Same as above.

File: token-syncer/src/token_syncer/commands/ping.clj (hunk 2 above)
Comment (shamsimam) [reply token-syncer/src/token_syncer/commands/ping.clj]: Changed to lower-case.

Comment (shamsimam) [reply token-syncer/src/token_syncer/commands/ping.clj]: Added.

File: token-syncer/src/token_syncer/waiter.clj (hunk 6 above)
Comment (shamsimam) [reply token-syncer/src/token_syncer/waiter.clj]: Nick started doing it and I'm following suit as I like it. It explicitly returns `nil` rather than us relying on the output of the `log/info` call to be `nil`.

File: token-syncer/test/token_syncer/waiter_test.clj (hunk 7 above)
Comment (shamsimam) [reply token-syncer/test/token_syncer/waiter_test.clj]: Oops sorry, this is a bug fix. The assertions were previously ignoring the `in-endopint-url` and treating it as a the message for a failed assertion.

File: token-syncer/src/token_syncer/waiter.clj (hunk 5 above)
Comment (shamsimam) [reply token-syncer/src/token_syncer/waiter.clj]: Done, added a `-q` flag to configure this in the sub-command.

File: token-syncer/src/token_syncer/commands/ping.clj (hunk 3 above)
Comment (dposada) [reply token-syncer/src/token_syncer/commands/ping.clj]: Got it

File: token-syncer/src/token_syncer/waiter.clj (hunk 5 above)
Comment (dposada) [reply token-syncer/src/token_syncer/waiter.clj]: Is there a long-form for it as well?

Comment (shamsimam) [reply token-syncer/src/token_syncer/waiter.clj]: Yes, `--queue-timeout`

File: token-syncer/src/token_syncer/waiter.clj (hunk 6 above)
Comment (shamsimam) [reply token-syncer/src/token_syncer/waiter.clj]: I changed it to explicitly return `nil`.

File: token-syncer/integration/token_syncer/basic_test.clj (hunk 1 above)
Comment (dposada) [reply token-syncer/integration/token_syncer/basic_test.clj]: A comment explaining this would be nice
